File: java-client/src/main/java/com/couchbase/client/java/bucket/Bucket.java
package com.couchbase.client.java.bucket;

import com.couchbase.client.java.query.Criteria;
import com.couchbase.client.java.query.QueryResult;
import com.couchbase.client.java.view.View;
import com.couchbase.client.java.view.ViewQueryOptions;
import com.couchbase.client.java.view.ViewResult;
import reactor.core.composable.Promise;

import java.util.List;

/**
 * Operations that can be performed against a bucket.
 */
public interface Bucket {

    Promise<Document> insert(Document document);
    Promise<Document> insert(String id, Object value);
    Promise<Document> insert(String id, Object value, int expiration);
    Promise<Document> insert(String id, Object value, int expiration, String groupId);
    Promise<List<Document>> insert(Iterable<Document> documents);

    Promise<Document> upsert(Document document);
    Promise<Document> upsert(String id, Object document);
    Promise<Document> upsert(String id, Object document, int expiration);
    Promise<Document> upsert(String id, Object document, int expiration, String groupId);
    Promise<List<Document>> upsert(Iterable<Document> documents);

    Promise<Document> replace(Document document);
    Promise<Document> replace(String id, Object document);
    Promise<Document> replace(String id, Object document, long cas);
    Promise<Document> replace(String id, Object document, long cas, int expiration);
    Promise<Document> replace(String id, Object document, long cas, int expiration, String groupId);
    Promise<List<Document>> replace(Iterable<Document> documents);

    Promise<Document> update(Document document);
    Promise<Document> update(String id, Object document);
    Promise<Document> update(String id, Object document, long cas);
    Promise<Document> update(String id, Object document, long cas, int expiration);
    Promise<Document> update(String id, Object document, long cas, int expiration, String groupId);
    Promise<List<Document>> update(Iterable<Document> documents);

    Promise<Document> remove(String id);
    Promise<Document> remove(String id, long cas);
    Promise<Document> remove(String id, long cas, String groupId);
    Promise<List<Document>> remove(Iterable<Document> documents);

    Promise<Document> get(String id);
    Promise<Document> get(String id, int lock);
    Promise<Document> get(String id, int lock, String groupId);
    Promise<Document> get(Document document);
    Promise<Document> get(Document document, int lock);
    Promise<List<Document>> get(Iterable<Document> documents);
    Promise<List<Document>> get(Iterable<Document> documents, int lock);

    Promise<List<Document>> find(View view);
    Promise<List<Document>> find(Criteria criteria);

    Promise<Document> unlock(Document document);
    Promise<Document> unlock(String id, long cas);
    Promise<Document> unlock(String id, long cas, String groupId);

    Promise<Document> counter(Document document);
    Promise<Document> counter(Document document, int delta);
    Promise<Document> counter(Document document, int delta, int initial);
    Promise<Document> counter(String id, int delta);
    Promise<Document> counter(String id, int delta, int initial);
    Promise<Document> counter(String id, int delta, int initial, int expiration);
    Promise<Document> counter(String id, int delta, int initial, int expiration, String groupId);

    Promise<ViewResult> view(String designDocument, String viewName, ViewQueryOptions options);
    Promise<QueryResult> query(String query);

    void endure(PersistTo persistTo, ReplicateTo replicateTo, BucketCallback callback);

    Promise<Boolean> flush();
    Promise<BucketInfo> info();

    Promise<DesignDocument> insertDesignDocument(DesignDocument designDocument, boolean overwrite);
    Promise<DesignDocument> insertDesignDocument(String name, String content, boolean overwrite);
    Promise<DesignDocument> updateDesignDocument(DesignDocument designDocument);
    Promise<DesignDocument> updateDesignDocument(String name, String content);
    Promise<DesignDocument> removeDesignDocument(String name);
    Promise<DesignDocument> removeDesignDocument(DesignDocument designDocument);
    Promise<DesignDocument> getDesignDocument(String name);
    Promise<List<DesignDocument>> listDesignDocuments();
}

File: scala-client/src/main/scala/com/couchbase/client/scala/bucket/Bucket.scala
package com.couchbase.client.scala.bucket

import scala.concurrent.Future
import com.couchbase.client.scala.view.{ViewQueryOptions, View, ViewResult}
import com.couchbase.client.scala.query.{QueryResult, Criteria}

/**
 * Operations that can be performed against a bucket.
 */
trait Bucket {

  def insert(document: Document): Future[Document]
  def insert(id: String, value: AnyRef, expiration: Int = Unit, groupId: String = Unit): Future[Document]
  def insert(documents: Iterable[Document]): Future[List[Document]]

  def upsert(document: Document): Future[Document]
  def upsert(id: String, value: AnyRef, expiration: Int = Unit, groupId: String = Unit): Future[Document]
  def upsert(documents: Iterable[Document]): Future[List[Document]]

  def replace(document: Document): Future[Document]
  def replace(id: String, value: AnyRef, cas: Long = Unit, expiration: Int = Unit, groupId: String = Unit): Future[Document]
  def replace(documents: Iterable[Document]): Future[List[Document]]

  def update(document: Document): Future[Document]
  def update(id: String, value: AnyRef, cas: Long = Unit, expiration: Int = Unit, groupId: String = Unit): Future[Document]
  def update(documents: Iterable[Document]): Future[List[Document]]

  def remove(document: Document): Future[Document]
  def remove(id: String, cas: Long = Unit, groupId: String = Unit): Future[Document]
  def remove(documents: Iterable[Document]): Future[List[Document]]

  def get(id: String, lock: Int = Unit, groupId: String = Unit): Future[Document]
  def get(document: Document, lock: Int = Unit): Future[Document]
  def get(documents: Iterable[Document], lock: Int = Unit): Future[List[Document]]

  def endure(persistTo: PersistTo, replicateTo: ReplicateTo, callback: Bucket => Unit): Unit

  def find(view: View): Future[List[Document]]
  def find(criteria: Criteria): Future[List[Document]]

  def unlock(document: Document): Future[Document]
  def unlock(id: String, cas: Long, groupId: String = Unit): Future[Document]

  def counter(document: Document, delta: Int, initial: Int = Unit): Future[Document]
  def counter(id: String, delta: Int, initial: Int = Unit, expiration: Int = Unit, groupId: String = Unit): Future[Document]

  def view(designDocument: String, viewName: String, options: ViewQueryOptions = Unit): Future[ViewResult]
  def query(query: String): Future[QueryResult]

  def flush: Future[Boolean]
  def info: Future[BucketInfo]

  def insertDesignDocument(designDocument: DesignDocument, overwrite: Boolean = true): Future[DesignDocument]
  def insertDesignDocument(name: String, content: String, overwrite: Boolean = true): Future[DesignDocument]
  def updateDesignDocument(designDocument: DesignDocument): Future[DesignDocument]
  def updateDesignDocument(name: String, content: String): Future[DesignDocument]
  def removeDesignDocument(name: String): Future[DesignDocument]
  def removeDesignDocument(designDocument: DesignDocument): Future[DesignDocument]
  def getDesignDocument(name: String): Future[DesignDocument]
  def listDesignDocuments: Future[List[DesignDocument]]


}

